java/monkey/src/main/java/root/Parser.java

```diff
@@ -0,0 +1,104 @@
+package root;
+
+import root.ast.*;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class Parser {
+
+    private final Lexer lexer;
+
+    private Token currentToken;
+
+    private Token peekToken;
+
+    public final List<String> errors = new ArrayList<>();
+
+    public Parser(Lexer lexer) {
+        this.lexer = lexer;
+
+        this.nextToken();
+        this.nextToken();
+    }
+
+    public Program parseProgram() {
+        Program program = new Program();
+
+        while (!curTokenIs(TokenType.EOF)) {
+            parseStatement().ifPresent(program.statements::add);
+            nextToken();;
+        }
+
+        return program;
+    }
+
+    private void peekError(TokenType expected) {
+        errors.add("Expected next token to be %s, got %s".formatted(expected.name(), currentToken.type().name()));
+    }
+
+    private void nextToken() {
+        this.currentToken = this.peekToken;
+        this.peekToken = this.lexer.nextToken();
+    }
+
+    private Optional<Statement> parseStatement() {
+        return switch (this.currentToken.type()) {
+            case LET -> this.parseLetStatement();
+            case RETURN -> this.parseReturnStatement();
+            default -> Optional.empty();
+        };
+    }
+
+    private Optional<Statement> parseReturnStatement() {
+        ReturnStatement returnStatement = new ReturnStatement(this.currentToken);
+
+        this.nextToken();
+
+        // TODO: We're skipping the expressions until we encounter a semicolon
+        while (!this.curTokenIs(TokenType.SEMI)) {
+            this.nextToken();
+        }
+
+        return Optional.of(returnStatement);
+    }
+
+    private Optional<Statement> parseLetStatement() {
+        LetStatement letStatement = new LetStatement(this.currentToken);
+
+        if (!this.expectPeek(TokenType.IDENT)) {
+            return Optional.empty();
```

Comment:
Multiple return statements in a function is a code smell. Can we flatten this? Pattern Matching might help.

Reply:
Hmm, I don't really know how to make this function have a single return statement while keeping things readable and not repeating code, specially because the `expectPeek` has side effects. I can think of a way to do it using optionals but I don't think you'll like it:

```java
   private Optional<Statement> parseLetStatement() {
        LetStatement letStatement = new LetStatement(this.currentToken);

        return this.expectPeek(TokenType.IDENT).flatMap(identToken -> {
            letStatement.name = new Identifier(identToken, identToken.literal());

            return this.expectPeek(TokenType.ASSIGN);
        }).map((assignToken -> {

            // TODO: We're skipping the expressions until we encounter a semicolon
            while (!this.curTokenIs(TokenType.SEMI)) {
                this.nextToken();
            }

            return letStatement;
        }));
    }
```

Changing the `expectPeek` function to: 

```java
    private Optional<Token> expectPeek(TokenType type) {
        if (this.peekTokenIs(type)) {
            this.nextToken();
            return Optional.of(this.currentToken);
        }

        this.peekError(type);
        return Optional.empty();
    }
```

If you know another way to write this please let me know

Reply:
Also, I don't really see the original implementation as having multiple return statements since the only valid return is the last one at the end. The other are merely error states that when encountered means we should stop building the Statement/Expression

Reply:
If they are errors, we should throw Exceptions.

Status Errors can be ignored, Exceptions cannot.

Reply:
Many interpreters and compilers stop on the first error and only report on it, but I believe the idea in the book is to output all errors on a single run so the programmer can fix them all in one go, that's why I don't believe exceptions to be the best idea here (I also don't like the ideas of exceptions on general because of the hidden and disruptive control flow and all problems that come with it, and they should be only be thrown on truly unrecoverable errors or when the programmer misuses the api, but I know that's an unpopular opinion among java programmers).

I changed the implementation to a more java-esque one with no Optional. I guess my time on Rust is getting to me lol

Reply:
Yeah no optional and instead returning the `null` is best if you want performance.
Exceptions aren't bad (in general) as long as you don't overboard with them.